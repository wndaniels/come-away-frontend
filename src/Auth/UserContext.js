import React from "react";
/**
 *  Context:
 *
 * Provides currentUser object and setter for it throughout app.
 */
const UserContext = React.createContext();

export default UserContext;
