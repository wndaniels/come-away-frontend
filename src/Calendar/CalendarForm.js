import React, { useContext, useEffect, useState } from "react";
import useLocalStorage from "../hooks/useLocalStorage.js";
import { useNavigate } from "react-router-dom";
import ComeAwayApi from "../Api/api.js";
import UserContext from "../Auth/UserContext.js";
import jwt from "jsonwebtoken";
import Alert from "../Common/Alert.js";

export const TOKEN_STORAGE_ID = "comeaway-token";

const CalendarForm = () => {
  const navigate = useNavigate();
  const { currentUser, setCurrentUser } = useContext(UserContext);
  const [infoLoaded, setInfoLoaded] = useState(false);
  const [token, setToken] = useLocalStorage(TOKEN_STORAGE_ID);
  const [calAvailBegin, setCalAvailBegin] = useState([]);
  const [calAvailEnd, setCalAvailEnd] = useState([]);
  const [formData, setFormData] = useState({
    businessBeginsHour: 0,
    businessEndsHour: 0,
    userId: currentUser.id,
  });

  const [formError, setFormError] = useState([]);

  useEffect(
    function loadUserInfo() {
      async function getCurrentUser() {
        if (token) {
          try {
            let { username } = jwt.decode(token);
            ComeAwayApi.token = token;
            let currentUser = await ComeAwayApi.getCurrentUser(username);
            setCurrentUser(currentUser);
          } catch (errors) {
            setFormError(errors);
            setCurrentUser(null);
          }
        }
        setInfoLoaded(true);
      }
      setInfoLoaded(false);
      getCurrentUser();
    },
    [token, setCurrentUser]
  );

  useEffect(function getCalData() {
    async function getBeginHoursData() {
      const beginHourRes = await ComeAwayApi.getBeginHours();
      setCalAvailBegin(beginHourRes);
    }

    async function getEndHourData() {
      const endHourRes = await ComeAwayApi.getEndHours();
      setCalAvailEnd(endHourRes);
    }

    getBeginHoursData([]);
    getEndHourData([]);
  }, []);

  async function handleSubmit(evt) {
    evt.preventDefault();

    let calendarData = {
      businessBeginsHour: formData.businessBeginsHour,
      businessEndsHour: formData.businessEndsHour,
      userId: currentUser.id,
    };

    let username = currentUser.username;

    let createdCal;

    try {
      createdCal = await ComeAwayApi.createCal(username, calendarData);
    } catch (errors) {
      setFormError([errors]);
      return;
    }

    setFormData((f) => ({ ...f }));
    setFormError([]);
    navigate("/");
  }

  function handleChange(evt) {
    const { name, value } = evt.target;

    setFormData((f) => ({
      ...f,
      [name]: parseInt(value),
    }));

    setFormError([]);
  }

  if (!infoLoaded) return;

  return (
    <div className="CalendarForm mt-5">
      <div className="container col-md-6 offset-md-3 col-lg-4 offset-lg-4">
        <h1 className="mb-3">Create Calendar</h1>
        <div className="card">
          <div className="card-body">
            <form method="post">
              <div className="d-grid gap-3">
                {formError.length ? (
                  <Alert
                    type="danger"
                    messages={["All fields must be complete."]}
                  ></Alert>
                ) : null}

                <div className="form-group">
                  <label htmlFor="businessBeginsHour">
                    * Select Start Time:
                  </label>
                  <select
                    name="businessBeginsHour"
                    className="form-control"
                    onChange={handleChange}
                  >
                    <option hidden>-</option>
                    {calAvailBegin &&
                      calAvailBegin.map((s, id) => (
                        <option value={s.businessBeginsHour} key={id}>
                          {s.hourTitle}
                        </option>
                      ))}
                  </select>
                </div>

                <div className="form-group">
                  <label htmlFor="businessEndsHour">* Select End Time:</label>
                  <select
                    name="businessEndsHour"
                    className="form-control mb-3"
                    onChange={handleChange}
                  >
                    <option hidden>-</option>
                    {calAvailEnd &&
                      calAvailEnd.map((e, id) => (
                        <option value={e.businessEndsHour} key={id}>
                          {e.hourTitle}
                        </option>
                      ))}
                  </select>
                </div>
              </div>

              <button onClick={handleSubmit} className="btn btn-sm btn-primary">
                Create Calendar
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
};

export default CalendarForm;
