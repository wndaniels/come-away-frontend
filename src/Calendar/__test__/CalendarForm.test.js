import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import "@testing-library/jest-dom";
import CalendarForm from "../CalendarForm";
import UserContext from "../../Auth/UserContext";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => jest.fn(),
}));

describe("CalendarForm", () => {
  it("renders without crashing", () => {
    const { asFragment } = render(
      <Wrapper>
        <CalendarForm />
      </Wrapper>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("can find placeholder", () => {
    const { getByText } = render(
      <Wrapper>
        <CalendarForm />
      </Wrapper>
    );
    const startTimeSelect = getByText(/select start time:/i);
    expect(startTimeSelect).toBeInTheDocument();
  });
});

function Wrapper({ children }) {
  return (
    <UserContext.Provider
      value={{
        currentUser: jest.fn(),
        setCurrentUser: jest.fn(),
        login: jest.fn(),
        signup: jest.fn(),
        logout: jest.fn(),
      }}
    >
      {children}
    </UserContext.Provider>
  );
}
