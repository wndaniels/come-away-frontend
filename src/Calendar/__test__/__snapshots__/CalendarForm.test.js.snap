// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarForm renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="CalendarForm mt-5"
  >
    <div
      class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
    >
      <h1
        class="mb-3"
      >
        Create Calendar
      </h1>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <form
            method="post"
          >
            <div
              class="d-grid gap-3"
            >
              <div
                class="form-group"
              >
                <label
                  for="businessBeginsHour"
                >
                  * Select Start Time:
                </label>
                <select
                  class="form-control"
                  name="businessBeginsHour"
                >
                  <option
                    hidden=""
                  >
                    -
                  </option>
                </select>
              </div>
              <div
                class="form-group"
              >
                <label
                  for="businessEndsHour"
                >
                  * Select End Time:
                </label>
                <select
                  class="form-control mb-3"
                  name="businessEndsHour"
                >
                  <option
                    hidden=""
                  >
                    -
                  </option>
                </select>
              </div>
            </div>
            <button
              class="btn btn-sm btn-primary"
            >
              Create Calendar
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
